arch/xtensa/core/mpu.c

```diff
@@ -701,3 +707,419 @@ void xtensa_mpu_init(void)
+		/* Account for wrapping around back to 0. */
+		stack_end_addr = 0xFFFFFFFFU;
+	}
+
+	/*
+	 * Allow USER access to the thread's stack in its new domain if
+	 * we are migrating. If we are not migrating this is done in
+	 * xtensa_user_stack_perms().
+	 */
+	if (is_migration) {
+		/* Add stack to new domain's MPU map. */
+		ret = mpu_map_region_add(&domain->arch.mpu_map,
+					 thread->stack_info.start, stack_end_addr,
+					 XTENSA_MPU_ACCESS_P_RW_U_RW,
+					 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+					 NULL);
+
+		/* Probably this fails due to no more available slots in MPU map. */
+		__ASSERT_NO_MSG(ret == 0);
+	}
+
+	thread->arch.mpu_map = &domain->arch.mpu_map;
+
+	/*
+	 * Remove thread stack from old memory domain if we are
+	 * migrating away from old memory domain. This is done
+	 * by simply remove USER access from the region.
+	 */
+	if (is_migration) {
+		/*
+		 * Remove stack from old MPU map by...
+		 * "adding" a new memory region to the map
+		 * as this carves a hole in the existing map.
+		 */
+		ret = mpu_map_region_add(old_map,
+					 thread->stack_info.start, stack_end_addr,
+					 XTENSA_MPU_ACCESS_P_RW_U_NA,
+					 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+					 NULL);
+	}
+
+	/*
+	 * Need to switch to new MPU map if this is the current
+	 * running thread.
+	 */
+	if (thread == _current_cpu->current) {
+		xtensa_mpu_map_write(thread->arch.mpu_map);
+	}
+
+	return ret;
+}
+
+int arch_mem_domain_thread_remove(struct k_thread *thread)
+{
+	uintptr_t stack_end_addr;
+	int ret;
+
+	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
+
+	if ((thread->base.user_options & K_USER) == 0) {
+		ret = 0;
+		goto out;
+	}
+
+	if ((thread->base.thread_state & _THREAD_DEAD) == 0) {
+		/* Thread is migrating to another memory domain and not
+		 * exiting for good; we weren't called from
+		 * z_thread_abort().  Resetting the stack region will
+		 * take place in the forthcoming thread_add() call.
+		 */
+		ret = 0;
+		goto out;
+	}
+
+	stack_end_addr = thread->stack_info.start + thread->stack_info.size;
+	if (stack_end_addr < thread->stack_info.start) {
+		/* Account for wrapping around back to 0. */
+		stack_end_addr = 0xFFFFFFFFU;
+	}
+
+	/*
+	 * Restore permissions on the thread's stack area since it is no
+	 * longer a member of the domain.
+	 */
+	ret = mpu_map_region_add(&domain->arch.mpu_map,
+				 thread->stack_info.start, stack_end_addr,
+				 XTENSA_MPU_ACCESS_P_RW_U_NA,
+				 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+				 NULL);
+
+	xtensa_mpu_map_write(thread->arch.mpu_map);
+
+out:
+	return ret;
+}
+
+int arch_buffer_validate(void *addr, size_t size, int write)
+{
+	uintptr_t start_addr, end_addr;
+	struct xtensa_mpu_entry *entry_slot_s, *entry_slot_e;
+	bool exact_s, exact_e;
+	uint8_t idx_s, idx_e, first_enabled_idx;
+
+	int ret = 0;
+	const struct k_thread *thread = _current;
+	const struct xtensa_mpu_map *map = thread->arch.mpu_map;
+	const struct xtensa_mpu_entry *entries = map->entries;
```

Comment:
I don't understand why we're mucking with the MPU table directly here?  This is a request for the current thread's address space, can't you just write a loop doing a PPTLB every 4k?

Reply:
errr.... forgot about PPTLB. Updated to use that instead.